source/grid/manifold_lib.cc

```diff
@@ -280,67 +341,137 @@ get_new_point (const ArrayView<const Point<spacedim>> &vertices,
                const ArrayView<const double>          &weights) const
 {
   const unsigned int n_points = vertices.size();
+  const double tolerance = 1e-10;
 
-  double rho = 0.0;
+  double rho = 0.;
   Tensor<1,spacedim> candidate;
-  for (unsigned int i = 0; i<n_points; i++)
-    {
-      const Tensor<1,spacedim> direction (vertices[i]-center);
-      rho += direction.norm()*weights[i];
-      candidate += direction*weights[i];
-    }
-
-  // Unit norm direction.
-  candidate /= candidate.norm();
 
-  return center+rho*candidate;
-}
+  // Step 1:
+  // Perform a simple average ...
+  {
+    double total_weights = 0.;
+    for (unsigned int i = 0; i < n_points; i++)
+      {
+        if (std::abs(1-weights[i])<tolerance)
+          return vertices[i];
+        const Tensor<1, spacedim> direction(vertices[i] - center);
+        rho += direction.norm() * weights[i];
+        candidate += direction * weights[i];
+        total_weights += weights[i];
+      }
+    // ... and normalize.
+    candidate /= candidate.norm();
+    rho /= total_weights;
+  }
 
-// ============================================================
-// CylindricalManifold
-// ============================================================
+  if (spacedim<2)
+    return center + rho*candidate;
 
-namespace internal
-{
-  namespace CylindricalManifold
+  // Step 2:
+  // Do Newton-style iterations to improve the estimate.
+  //
+  // In this step, we consider all points and directions to be embedded
+  // in a three-dimensional space.
   {
-    namespace
-    {
-      // helper function to compute a vector orthogonal to a given one.
-      template <int spacedim>
-      Point<spacedim>
-      compute_normal(const Tensor<1,spacedim> &vector)
+    std::vector<Tensor<1, 3> > directions(vertices.size());
```

Comment:
Can you use `boost::container::small_vector<Tensor<1, 3>, 100> directions(n_points);` instead?